src/components/vm/overview/cpuModal.jsx

```diff
@@ -0,0 +1,326 @@
+        if (num % i === 0) {
+            divs.push(i);
+        }
+    }
+
+    if (num > 1) {
+        divs.push(num);
+    }
+
+    return divs;
+};
+
+const clamp = (value, max, min) => {
+    return value < min || isNaN(value) ? min : (value > max ? max : value);
+};
+
+export const CPUModal = ({ vm, maxVcpu, models }) => {
+    const Dialogs = useDialogs();
+
+    const [error, setError] = useState(undefined);
+    const [sockets, setSockets] = useState(vm.cpu.topology.sockets || 1);
+    const [threads, setThreads] = useState(vm.cpu.topology.threads || 1);
+    const [cores, setCores] = useState(vm.cpu.topology.cores || 1);
+    const [max, setMax] = useState(parseInt(vm.vcpus.max) || 1);
+    const [count, setCount] = useState(parseInt(vm.vcpus.count) || 1);
+    const [cpuMode, setCpuMode] = useState(vm.cpu.mode);
+    const [cpuModel, setCpuModel] = useState(vm.cpu.model);
+    const [isLoading, setIsLoading] = useState(false);
+
+    function onMaxChange(value) {
+        // Allow empty string
+        if (value == "") {
+            setMax("");
+            return;
+        }
+
+        const maxHypervisor = parseInt(maxVcpu);
+        let maxValue = parseInt(value);
+
+        // Check new value for limits
+        maxValue = clamp(maxValue, maxHypervisor, 1);
+
+        // Recalculate new values for sockets, cores and threads according to new max value
+        // Max value = Sockets * Cores * Threads
+        const state = { max: maxValue, sockets, cores };
+
+        // If count of used VCPU greater then new max value, then change it to new max value
+        if (maxValue < count) {
+            state.count = maxValue;
+        }
+
+        // Recalculate sockets first, and get array of all divisors of new max values
+        let divs = dividers(state.max);
+
+        // If current sockets value is not in divisors array, then change it to max divisor
+        if (divs.indexOf(sockets) === -1 || (vm.cpu.topology.sockets || 1) === sockets) {
+            state.sockets = divs[divs.length - 1];
+        }
+
+        // Get next divisors
+        divs = dividers(state.max / state.sockets);
+        if (divs.indexOf(cores) === -1) {
+            state.cores = divs[divs.length - 1];
+        }
+
+        // According to: Max value = Sockets * Cores * Threads. Threads = Max value / ( Sockets * Cores )
+        state.threads = state.max / (state.cores * state.sockets);
+        setSockets(state.sockets);
+        setCores(state.cores);
+        setThreads(state.threads);
+        if (state.count)
+            setCount(state.count);
+        setMax(state.max);
+    }
+
+    function onCountSelect (value) {
+        // Allow empty string
+        if (value == "") {
+            setCount("");
+            return;
+        }
+
+        const newValue = clamp(value, max, 1);
+        setCount(parseInt(newValue));
+    }
+
+    function onSocketChange (value) {
+        const state = { sockets, cores };
+        state.sockets = parseInt(value);
```

Comment:
This is an anti-practice - this is changing the component state without the set* function.

Reply:
But this is not the `state` itself, it's just a local variable, local object `const state = { sockets, cores };`, So changing `state.socket = something` doesn't change react state, only the local variable.
But I admit the name of variable is confusing... perhaps let's call it stateDelta, or newState?